lib/nexthop.c

```diff
@@ -316,12 +316,47 @@ unsigned int nexthop_level(struct nexthop *nexthop)
 
 uint32_t nexthop_hash(struct nexthop *nexthop)
 {
-	uint32_t key;
+	uint32_t key = 0x45afe398;
 
-	key = jhash_1word(nexthop->vrf_id, 0x45afe398);
-	key = jhash_1word(nexthop->ifindex, key);
-	key = jhash_1word(nexthop->type, key);
-	key = jhash(&nexthop->gate, sizeof(union g_addr), key);
+	key = jhash_3words(nexthop->type, nexthop->vrf_id,
+			   nexthop->nh_label_type, key);
+	/* gate and blackhole are together in a union */
+	key = jhash(&nexthop->gate, sizeof(nexthop->gate), key);
+	key = jhash(&nexthop->src, sizeof(nexthop->src), key);
+	key = jhash(&nexthop->rmap_src, sizeof(nexthop->rmap_src), key);
 
+	if (nexthop->nh_label) {
+		int i = nexthop->nh_label->num_labels;
+
+		while (i >= 3) {
+			key = jhash_3words(nexthop->nh_label->label[i],
+					   nexthop->nh_label->label[i + 1],
```

Comment:
we can't really use `i` this way, can we - aren't we looking beyond the end of the array? I think we need two variables - one counting forward from zero, and a second one counting down the remaining number of labels to process.

Reply:
Oh man you are 100% right, sorry I didn't even notice that in the suggestion.

Reply:
looks good now!